File: code/src/wang/relish/algorithm/handler/Looper.java
package wang.relish.algorithm.handler;

import java.util.ArrayList;
import java.util.List;

//循环读取者
public class Looper {

    public static ThreadLocal<Looper> looperThreadLocal;
    private static Looper mMainLooper;

    static {
        looperThreadLocal = new ThreadLocal<>();
    }

    private Message mMessage;
    private MessageQueue mMessageQueue = new MessageQueue();
    private volatile boolean quit = false;

    private Looper() {
    }

    public static Looper getLooper() {
        return looperThreadLocal.get();
    }

    public static Looper getMainLooper() {
        return mMainLooper;
    }

    public static void prepare() {
        if (looperThreadLocal.get() != null) {
            throw new Error("current thread already has prepared!");
        }

        looperThreadLocal.set(new Looper());
    }

    public static void prepareMain() {
        prepare();
        mMainLooper = getLooper();
    }

    public static void loop() {
        //do something
        Looper looper = looperThreadLocal.get();
        if (looper == null) {
            throw new Error("please use prepare() to create looper!");
        }

        while (true) {

            looper.mMessage = looper.mMessageQueue.peek();
            if (looper.mMessage == null /*|| !looper.mMessage.doHandleMessage()*/) {
                continue;
            }

            //开始处理消息
            looper.mMessage = looper.mMessageQueue.poll();
            looper.mMessage.target.handleMessage(looper.mMessage);
            //处理完消息后，放入缓存池
//            Message.addCache(looper.mMessage);
            looper.mMessage = null;
            if (looper.quit) return;
        }
    }

    public void quit() {
        this.quit = true;
    }

    public MessageQueue getMessageQueue() {
        return mMessageQueue;
    }
}

File: code/src/wang/relish/algorithm/handler/mini/Looper.kt
package wang.relish.algorithm.handler.mini

class Looper {

    internal val mQueue: MessageQueue
    private var quit: Boolean = false

    private constructor() {
        mQueue = MessageQueue()
    }

    fun quit() {
        quit = true
    }

    companion object {
        @JvmStatic
        val sThreadLocal = ThreadLocal<Looper>()

        @JvmStatic
        lateinit var sMainLooper: Looper

        @JvmStatic
        fun myLooper(): Looper = sThreadLocal.get()

        @JvmStatic
        fun prepare() {
            if (sThreadLocal.get() != null) {
                throw RuntimeException("Only one Looper may be created per thread");
            }
            sThreadLocal.set(Looper())
        }

        @JvmStatic
        fun prepareMain() {
            prepare()
            sMainLooper = myLooper()
        }

        @JvmStatic
        fun getMainLooper(): Looper = sMainLooper

        @JvmStatic
        fun loop() {
            val me = myLooper()
            if (me == null) {
                throw RuntimeException("No Looper; Looper.prepare() wasn't called on this thread.");
            }
            while (true) {
                val msg = me.mQueue.next()
                if (msg == null) {
                    continue
                }
                msg.target?.dispatchMessage(msg)
                if (me.quit) return
            }
        }
    }
}